### PR TITLE
FEniCSx: Updates for 0.5.1

### PR DESCRIPTION
**Pull request by jhale**
This (in my view!) supersedes https://github.com/spack/spack/pull/32611

Main differences:

- I have opted to deprecated some older versions. As FEniCSx is currently in a development phase there is little use keeping the old versions around with the additional complexity required in the Spack files.
- I have added the patch releases of Basix and DOLFINx.
- Looking forward to changes on `main` I have bumped cmake to `3.19:`. DOLFINx also does not require xtensor from now on.
- Basix also requires C++20.

This is a draft pending successful CI run on our repository.

### CODE REVIEW

**Comment by tldahlgren** [comment]
> This (in my view!) supersedes #32611

Oops.  I merged that without looking at this PR first.

**Review by chrisrichardson — DISMISSED** [review]

**Comment by jhale** [comment]
Please hold off on merging this until we've unmarked it as draft, there are a number of issues we'd like to fix.

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py
Comment (garth-wells) [inline var/spack/repos/builtin/packages/fenics-dolfinx/package.py:70]: `xtl` dependency should be kept since its header files are included directly.

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py
Comment (garth-wells) [inline var/spack/repos/builtin/packages/py-fenics-basix/package.py]: I don't think we want to make `numba` a variant. Users wanting `numba` can just concretize `py-fenics-basix` and `numba` together.

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py
Comment (garth-wells) [inline var/spack/repos/builtin/packages/py-fenics-basix/package.py]: Remove  - see above comment.

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py
Comment (garth-wells) [inline var/spack/repos/builtin/packages/py-fenics-basix/package.py]: Not needed since it's in the `PythonPackage` base class, see https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#dependencies.

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py
Comment (garth-wells) [inline var/spack/repos/builtin/packages/py-fenics-basix/package.py:37]: Add back since `xtl` is included directly.

File: var/spack/repos/builtin/packages/py-fenics-dolfinx/package.py
Comment (garth-wells) [inline var/spack/repos/builtin/packages/py-fenics-dolfinx/package.py]: Remove - see above comments.

File: var/spack/repos/builtin/packages/py-fenics-dolfinx/package.py
Comment (garth-wells) [inline var/spack/repos/builtin/packages/py-fenics-dolfinx/package.py]: Remove - see comment above.

File: var/spack/repos/builtin/packages/py-fenics-dolfinx/package.py
Comment (garth-wells) [inline var/spack/repos/builtin/packages/py-fenics-dolfinx/package.py]:
```suggestion
    depends_on("xtensor@0.23.10:", type="build", when="@:0.5")
```

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py
Comment (garth-wells) [inline var/spack/repos/builtin/packages/py-fenics-basix/package.py]: See comment above on pip.

**Review by garth-wells — CHANGES_REQUESTED** [review]

**Comment by jhale** [comment]
Thanks @garth-wells really helpful. I'll make the changes and then organise having this merged.

**Comment by jhale** [comment]
@chrisrichardson @garth-wells Could one of you give this an approval for an automatic merge?

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py
Comment (chrisrichardson) [inline var/spack/repos/builtin/packages/fenics-dolfinx/package.py]: Should `slepc` be "on" by default?

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/fenics-dolfinx/package.py]: This is not the correct sha256.  According to `spack checksum` it should be:

```suggestion
        sha256="e6462fc3b9653d335c28096f9b0393f966c55a64c77ed64cc6c949406cb1f2c3",
```

File: var/spack/repos/builtin/packages/py-fenics-dolfinx/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-fenics-dolfinx/package.py]: This also appears to be incorrect and should be:

```suggestion
        sha256="e6462fc3b9653d335c28096f9b0393f966c55a64c77ed64cc6c949406cb1f2c3",
```

File: var/spack/repos/builtin/packages/py-fenics-ffcx/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-fenics-ffcx/package.py]: This also does not match the output from `spack checksum`, which says it should be:

```suggestion
        sha256="f1dcd0973980706aba145274aeddeb771d2d730efcdb9b4be10adbe964f40d90",
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]
All but 3 new sha256 are confirmed.  The others are provided in suggestions.

**Comment by tldahlgren** [comment]
@spackbot run pipeline

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py
Comment (garth-wells) [reply var/spack/repos/builtin/packages/fenics-dolfinx/package.py]: I think on by default is ok.

Comment (jhale) [reply var/spack/repos/builtin/packages/fenics-dolfinx/package.py]: I'm not sure I agree @garth-wells. I would be more inclined to make the defaults as minimal as possible.

**Comment by jhale** [comment]
> All but 3 new sha256 are confirmed. The others are provided in suggestions.

Thanks for this detailed review work @tldahlgren, much appreciated.

**Review by chrisrichardson — DISMISSED** [review]

**Review by garth-wells — DISMISSED** [review]

**Review by tldahlgren — APPROVED** [review]
Confirmed all (19) new sha256.

**Comment by tldahlgren** [comment]
@chrisrichardson @garth-wells Did you want to take another look before this PR is merged?

**Comment by chrisrichardson** [comment]
Looks good

**Pull request merged**
